### ISSUE TITLE
False positive reports for F821 lint rule

### ISSUE THREAD

**Issue by chenzhekl** [astral-sh/ruff#17386]
### Summary

When we use Ruff with jaxtyping, Ruff falsely report F821 error for the following type annotation:

```python
import torch
from jaxtyping import Float, jaxtyped
from beartype import beartype

@jaxtyped(typechecker=beartype)
def foo(a: Float[torch.Tensor, "dim"]) -> None:
    pass
```

```
Undefined name `dim` Ruff[F821](https://docs.astral.sh/ruff/rules/undefined-name)
```

### Version

ruff 0.11.2

**Comment by Daverball** [comment]
A (admittedly kind of silly) workaround in the meantime would be to instead write it like this:

```python
import torch
from jaxtyping import Float, jaxtyped
from beartype import beartype

dim = "dim"

@jaxtyped(typechecker=beartype)
def foo(a: Float[torch.Tensor, dim]) -> None:
    pass
```

**Comment by ntBre** [comment]
Thanks for the report, and thanks @Daverball for the explanation.

I found #6014, which at least sounds related, but maybe not closely enough to call this a duplicate since it's for a different `typing` re-export.

**Comment by Orionnss** [comment]
An awful workaround that seems to work, you can add a whitespace before the name of the dimension as in 
`labels: Int[torch.Tensor, " nb_sentences"],` and it seems to work.

If someone is able to explain what this works, I would be very happy :)

**Comment by Daverball** [comment]
> If someone is able to explain why this works, I would be very happy :)

It's possible the annotations parser fails or gets skipped if the annotation starts with a whitespace, although I'd consider that a bug, that should get fixed, it's still a valid expression with the whitespace.

**Comment by Daverball** [comment]
Also #9860 is technically related, I only just saw this issue pop up as a duplicate recently. Maybe this issue should be closed in favor of that issue, although to be fair re-exporting heavily special cased type forms like `typing.Annotated` and `typing.Literal` under different names is a more general problem that isn't specific to jaxtyping, any library can do that and cause the same problems.

**Comment by knyazer** [comment]
By the way, this is also an issue (or a very similar issue it seems) in the original pyflakes :) https://github.com/PyCQA/pyflakes/issues/789

Maybe as a workaround we could split F821 into two rules, F821a and F821, where F821a refers to annotations? So that this rule can be disabled on its own. WDYT?

Also, another rule that is somewhat related is F722: in jaxtyping, the annotations of kind `Float[Array, ""]` are common, which triggers this rule, for, I presume, the same reason of using `Annotated as`.

**Comment by erlebach** [comment]
I'd like to know why this issue is so difficult to fix given the level of sophistication of so many other tools.

**Comment by patrick-kidger** [comment]
Have just come across this. As per the [jaxtyping FAQ](https://docs.kidger.site/jaxtyping/faq/#flake8-or-ruff-are-throwing-an-error) then adding a space at the start of the string -- `Float[Array, " foo"]` -- is my recommended work around. This converts the F821 error into an ignorable F722 error instead.

**Comment by kalk-ak** [comment]
Came here because I was facing the same issue
